### PR TITLE
thumbnail: compute I;16 rescale percentiles via histogram

### PR DESCRIPTION
**Pull request by sir-sigurd**
# Description

`_rescale_uint16_to_uint8` (the I;16 → 8-bit contrast stretch) calls `np.percentile` on the **full-resolution** array, which partitions a flattened copy of every pixel — transient memory that scales with image size. OOM kills are the lambda's documented failure mode, and this percentile copy is an image-sized *transient* allocated on top of the unavoidable output — removing it lowers the peak (other O(N) allocations like the LUT-indexed output remain).

uint16 has only 65536 possible values, so the percentiles can be read off a histogram instead. **Caveat (measured):** a single `np.bincount` over the whole array upcasts it to int64 (8 bytes/pixel) and costs *more* than `np.percentile`'s ~2 bytes/pixel uint16 copy — so the histogram is accumulated **per block** (1M-pixel blocks into a fixed 65536-bin int64 accumulator), keeping the int64 transient block-sized. Peak then stays bounded for any shape — the histogram is accumulated by chunking a flat iterator (`arr.flat[a:b]`), which copies only the requested slice regardless of contiguity, so even a degenerate `(1, N)` shape stays block-sized rather than costing more than the `np.percentile` baseline:

| approach | transient peak (72 MB uint16 array) |
|---|---|
| `np.percentile` | 74 MB (~2 B/px flatten copy, scales with N) |
| single `bincount` | 289 MB (~8 B/px int64 upcast) — rejected |
| **block-wise (this PR)** | **13 MB (bounded by block size)** |

Chunking the flat iterator (rather than `reshape(-1)`) is what keeps the non-contiguous color path (`arr[..., :3]`) bounded too — a `reshape(-1)` would copy that whole slice up front (~12 MB vs `np.percentile`'s ~70 MB on a 73 MB slice). As a side benefit it's also modestly faster (~1.5–2× on greyscale, ~1.4× on color: one `bincount` pass vs an introselect partition).

The `_percentile_uint16` helper tracks `np.percentile`'s default `"linear"` method (virtual index `q/100·(n-1)`, interpolate between the sorted values at floor/ceil, read from the CDF). Output matches `np.percentile` to floating-point tolerance — not bit-exact (numpy's `_lerp` is asymmetric for fraction ≥ 0.5), but the gap is sub-ULP and vanishes in the uint8 rescale, so the rescaled output and all checked-in thumbnails are **unchanged** (no golden churn; verified against the real I;16 fixtures plus synthetic fractional/constant/sparse/color cases). A unit test pins the equivalence.

(Credit: the single-`bincount` first cut had it backwards — it raised peak memory; caught in review and corrected to block-wise.)

# TODO

- [x] Unit tests — `test_percentile_uint16_matches_numpy` plus the existing `_convert_I16_to_L` / `_rescale` cases (output unchanged)
- [x] Changelog entry

<h3>Greptile Summary</h3>

This PR replaces `np.percentile` in `_rescale_uint16_to_uint8` with a block-wise histogram approach (`_percentile_uint16`) that computes the same percentiles without allocating an image-sized transient copy, reducing peak memory from O(N) to a fixed ~8 MB block budget.

- Histogram is accumulated via `arr.flat[a:b]`-chunked `np.bincount` calls into a fixed 65536-bin int64 accumulator, correctly handling both contiguous greyscale and non-contiguous color (`arr[..., :3]`) arrays.
- Percentile extraction replicates NumPy's `"linear"` method via CDF `searchsorted`, matching `np.percentile` output to sub-ULP tolerance so no golden thumbnails change.
- Two new test functions cover single-block equivalence (multiple distributions including skewed and color), multi-block correctness (via `monkeypatch` on `_HIST_BLOCK`), non-contiguous slices, and degenerate (1×N, N×1) shapes.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the change is a drop-in memory optimisation with output identical to the prior code, guarded by tight numerical tests.

The histogram accumulation, CDF lookup, and linear interpolation are all mathematically correct (verified by tracing through edge cases: n=1, repeated values, non-contiguous slices). The block-wise chunking via arr.flat[a:b] correctly bounds transient allocation regardless of array contiguity or shape. Test coverage is thorough: both single-block equivalence and forced multi-block accumulation paths are exercised, along with the non-contiguous RGBA slice and degenerate 1×N shapes that could otherwise have regressed silently.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| lambdas/thumbnail/src/t4_lambda_thumbnail/__init__.py | Adds _percentile_uint16 (block-wise histogram percentile) and wires it into _rescale_uint16_to_uint8; logic is correct and well-documented. |
| lambdas/thumbnail/tests/test_thumbnail.py | Adds two parametrized test functions covering numpy equivalence, multi-block accumulation, non-contiguous slices, and degenerate array shapes. |
| lambdas/thumbnail/CHANGELOG.md | Changelog entry added with correct PR link (#4968). |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["_rescale_uint16_to_uint8(arr)"] -->|arr.size == 0| B["return arr.astype(uint8)"]
    A -->|arr.size > 0| C["_percentile_uint16(arr, (0.01, 99.99))"]
    C --> D["counts = zeros(65536, int64)"]
    D --> E["flat = arr.flat"]
    E --> F{"for start in range(0, arr.size, 1M)"}
    F -->|each block| G["block = flat[start:start+1M]\ncounts += bincount(block, minlength=65536)"]
    G --> F
    F -->|done| H["cdf = cumsum(counts) / n = cdf[-1]"]
    H --> I{"for each q in qs"}
    I --> J["v = q/100*(n-1), lo_i = floor(v)\nval_lo = searchsorted(cdf, lo_i, right)\nval_hi = searchsorted(cdf, lo_i+1, right)\nout = val_lo + (val_hi-val_lo)*(v-lo_i)"]
    J --> I
    I -->|done| K["return [lo, hi]"]
    K --> L{"hi == lo?"}
    L -->|yes| M["fallback: min/max or constant"]
    L -->|no| N["LUT rescale → uint8"]
```

<sub>Reviews (3): Last reviewed commit: ["thumbnail: reframe the flat-chunk ration..."](https://github.com/quiltdata/quilt/commit/0102e4d249a4389f15e3af1b5b9e50b7bb1cd619) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36948460)</sub>

### CODE REVIEW

**Comment by sir-sigurd** [comment]
@greptileai

**Review by sir-sigurd — COMMENTED** [review]
Notes from an AI-assisted self-review pass:

The histogram swap produces identical output (tests pass), but it works against its own stated goal: on the uint16 arrays this path handles, it *raises* the percentile step's peak transient memory rather than lowering it. Holding line-level notes until the approach settles.

## Counterproductive on memory (the PR's stated goal)

The description assumes `np.percentile` allocates a float64 copy of every pixel (8 bytes/px). For uint16 input it doesn't: with the default `axis=None, overwrite_input=False` it takes the `arr = a.flatten()` branch, copying at the **input dtype** — uint16, 2 bytes/px — and partitions that in place. Only the two extracted percentile values ever become float64.

`np.bincount` is the one that allocates 8 bytes/px: it casts its entire input to int64 (`PyArray_ContiguousFromAny(..., NPY_INTP, ...)`) before counting. The 65536-bin result is tiny; the cost is the int64 copy of the full image.

Measured peak transient (calling the real `_percentile_uint16`, 36M-pixel inputs):

| path | `np.percentile` | this PR |
|---|---|---|
| greyscale `(H, W)` | ~2 B/px | ~8 B/px |
| color `(H, W, 4)[..., :3]` (the alpha path) | ~2 B/px | ~8.5 B/px |

So the percentile step goes from ~1× to ~4× the input array's footprint — on the OOM-prone path this change was meant to relieve. `np.percentile` is already the lighter option here.

If the goal is to get *below* 2 B/px: accumulate the histogram over row-blocks (per-block `bincount` into a fixed 65536-bin int64 accumulator) so the int64 transient stays block-sized. A single `bincount` over the whole array can't, because of the input upcast. The existing equivalence test is worth keeping for that path too.

## Minor (description accuracy)

- "bit-identical to `np.percentile`" is really "matches within ~1e-9": numpy's `_lerp` switches to `b - (b - a)(1 - t)` for fraction ≥ 0.5 (monotonicity), which this doesn't replicate. The gap is sub-ULP and vanishes after the uint8 rounding — output is unchanged — but the percentile values aren't literally bit-identical.
- The "8 bytes/pixel" figure describes the new code (the int64 cast), not `np.percentile`.

Checked: output equivalence (tests green), the percentile math against numpy's linear method, empty / single-value / constant edge cases, and the contiguous vs non-contiguous color paths.

🤖 Drafted with Claude; findings vetted by the reviewer — sanity-check rather than take as authoritative.

Reviewed against 01113d28

**Review by sir-sigurd — COMMENTED** [review]
Notes from an AI-assisted self-review pass — follow-up to the earlier one; likely the last.

The block-wise rework settles the memory concern from the previous pass. Measured peak transient on a 72 MB greyscale uint16 array: ~12 MB here vs ~72 MB for `np.percentile` — the histogram is genuinely lighter on the dominant I;16 path now, and the earlier description inaccuracies (the "bit-identical" wording, the 8 B/px attribution) are corrected. It's also a CPU win: ~4–5× faster than `np.percentile` on greyscale and ~2.6× on the color path (one `bincount` pass vs an introselect partition), with no regression anywhere — the description and CHANGELOG mention only memory, so that's an unstated upside worth a line. Two smaller items below.

## P2 — "bounded regardless of image size" doesn't hold for the RGBA path

The color branch calls `rescale(arr[..., :3])` in `generate_thumbnail`, passing a non-contiguous slice. Inside `_percentile_uint16`, `flat = arr.reshape(-1)` can't return a view of a non-contiguous array, so it copies the whole slice at ~2 bytes/pixel — peak scales with image size on that path, not fixed by block size.

Measured on a ~74 MB `[..., :3]` slice: ~82 MB transient here vs ~73 MB for `np.percentile` — about 9 MB / 12% *higher* than the code being replaced, on the metric this change optimizes. The net is still positive (the greyscale path dominates and improves a lot), but the unqualified "peak stays bounded regardless of image size" in the description, the comment above `_HIST_BLOCK`, the `_percentile_uint16` docstring, and the CHANGELOG ("lowering peak memory on large images") overclaim for color. Either scope the claim to contiguous input, or keep the color path bounded too.

*Alternatives: would blocking on the outer axis — copying each row-block to contiguous before `bincount` — keep the color path bounded, instead of one full `reshape(-1)` copy?*

## P3 — multi-block accumulation isn't exercised by a test

`_HIST_BLOCK` is 1M pixels; every array in `test_percentile_uint16_matches_numpy` is well under that (largest 200×200), so the per-block loop always runs exactly once. The accumulation across blocks — the part this commit adds — has no coverage. (Checked by hand: it matches `np.percentile` at 4 and 4000 iterations.) A case with a monkeypatched small `_HIST_BLOCK`, or one array above 1M pixels, would pin it.

Checked: the percentile math vs numpy's linear method, multi-block accumulation, empty / constant / single-value edges (guarded by the caller), and the contiguous vs non-contiguous color paths; re-measured peak transient on both.

🤖 Drafted with Claude; findings vetted by the reviewer — sanity-check rather than take as authoritative.

Reviewed against f8e066f7

File: lambdas/thumbnail/src/t4_lambda_thumbnail/__init__.py [hunk 1]
```diff
@@ -400,6 +400,29 @@ def handle_image(*, path: str, size: tuple[int, int], thumbnail_format: str):
...
+    # np.percentile on the full-resolution array allocates a float64 partition
+    # copy of every pixel (8 bytes each) — an OOM risk on large images, the
+    # known failure mode of this lambda. uint16 has only 65536 possible values,
+    # so a fixed-size histogram yields the same percentiles in O(65536) memory
+    # regardless of image size. Replicates np.percentile's default "linear"
+    # method exactly (verified bit-identical), so the rescaled output is
+    # unchanged. Channels are pooled (arr is flattened), matching the joint
+    # range np.percentile computed over the whole array.
+    cdf = np.cumsum(np.bincount(arr.reshape(-1), minlength=65536))
+    n = cdf[-1]
```
[the remark this answers is not in the document]
Comment (sir-sigurd) [reply lambdas/thumbnail/src/t4_lambda_thumbnail/__init__.py:434]: Leaving the guards off: `_percentile_uint16` is a private helper with a single caller (`_rescale_uint16_to_uint8`), which already returns early on empty input, and `qs` is always the constant `(0.01, 99.99)` — so neither an empty array nor an out-of-range q can reach here in practice. Adding validation would be dead defensiveness.

**Review by sir-sigurd — COMMENTED** [review]
Notes from a fuller self-review pass (not the quick inline look from before) — follow-up; the earlier two items (color-path bound, missing multi-block test) are both resolved by `fcfbe816`, which switched to first-axis blocking and added `test_percentile_uint16_multi_block`. Likely the last pass.

What this pass confirmed: output matches `np.percentile` within sub-ULP across a wide sweep (20k tie-heavy + 2k random arrays, plus 5k checks of the actual 65536-entry LUT — zero uint8 bucket flips, so `atol=1e-9` has ~5 orders of magnitude of margin); peak transient is bounded on both real paths (greyscale ~12 MB vs ~68 MB, color slice ~14 MB vs ~70 MB); ~3–5× faster CPU. Three items, none blocking.

## P2 — the bound is height-independent, not size-independent; wide single rows defeat it

First-axis blocking caps the per-block transient by *rows per block*, but `rows` clamps to 1 once a single row has ≥ `_HIST_BLOCK` (1M) elements — then each block is one full row and the int64 `bincount` of it scales with row width. Measured on a `(2, 20_000_000)` uint16 array (80 MB): ~159 MB transient, ~2× the `np.percentile` it replaces — the same worse-than-baseline behavior that motivated dropping the single-`bincount` version. Real images are roughly 2D-balanced, so this needs a row ≥1M px wide (≥350k for color) — pathological, and not a regression for any normal image — but the comment and the description both say peak "stays fixed / bounded **regardless of image size**," which is unconditionally false. Either scope that claim to image height, or chunk within a row so the bound holds for any shape.

*Alternatives: would capping each block by element count (sub-chunking a too-wide row along the last axis) keep the transient bounded for any shape, instead of one-row-minimum blocking?*

## P3 — is the percentile actually the binding peak of `generate_thumbnail`?

The description calls this "the last spot in the function still allocating proportional to pixel count," but `lut.astype(np.uint8)[arr]` produces a full 1 B/px output right after, the RGBA path adds an `np.dstack`, and `_rescale_float_to_uint8` still calls `np.percentile` on a full float copy (floats can't use the 65536-bin trick). The change is a real win on the uint16 path (transient ~4→3 B/px, CPU 3–5×), but if overall peak RSS is set by the decode / LUT-index / dstack, the OOM headroom gained is smaller than the percentile numbers alone suggest. Worth confirming the percentile was the binding allocation, or softening the "last spot" wording.

## P3 — `_percentile_uint16` isn't safe standalone (caller-guarded today)

On an empty or zero-dim array the helper either raises `IndexError` (`arr[0]` for `(0, W)`) or returns garbage `[65536, 65536]` (`(N, 0)`), and it doesn't validate `q`. Not reachable now — `_rescale_uint16_to_uint8` guards `if not arr.size` and hard-codes `(0.01, 99.99)` — but the guard lives one frame up, easy to miss when adding a second caller. A one-line precondition comment, or an early size guard in the helper, closes it. (This is Copilot's empty-input note — valid as reuse safety, not a live bug.)

Bot comments: Copilot's "use exact equality instead of `allclose`" would break a correct test — the symmetric-vs-numpy-`_lerp` gap is real and unavoidable, and exact `==` fails on roughly 1 in 1000 random arrays — so not adopting. Greptile's `#????` placeholder is already filled (`#4968`).

Checked: the percentile math vs numpy's linear method, peak memory on both paths, CPU, edge/degenerate shapes, security (input-driven allocation — see P2), and description/CHANGELOG accuracy. Holding line-level polish until the wording/approach in P2 is settled.

🤖 Drafted with Claude; findings vetted by the reviewer — sanity-check rather than take as authoritative.

Reviewed against fcfbe816

**Comment by sir-sigurd** [comment]
@greptileai — please re-review; the implementation was reworked since the last pass (now flat-iterator histogram chunking, head `0102e4d2`).

**Review by sir-sigurd — COMMENTED** [review]
Notes from an AI-assisted self-review pass — follow-up to the earlier passes; likely the last.

The flat-iterator rework (`arr.flat[a:b]`) settles the wide-row concern from the previous pass: the histogram transient is now bounded for any shape. Measured peak transient (fresh process each, `ru_maxrss`): greyscale 72 MB array ~14 MB here vs ~69 MB for `np.percentile`; non-contiguous color slice ~12 MB vs ~70 MB; and the `(2, 20M)` wide row that defeated first-axis blocking ~12 MB vs ~77 MB (the previous approach hit ~159 MB on that shape). Output unchanged — the equivalence and rescale tests pass. Two description-accuracy items remain; the code is clean.

## P2 — the description contradicts itself (and the code) on the blocking mechanism

The "How" section says the histogram "is accumulated by chunking a flat iterator (`arr.flat[a:b]`)" — matching the code — but a later paragraph says "Blocking is over the **first axis** (not a flat view)". The code is `flat = arr.flat; flat[start:start + _HIST_BLOCK]`, so the "(not a flat view)" parenthetical is false against HEAD; it's leftover from the earlier first-axis version (`fcfbe816`). The reasoning in that paragraph (don't `reshape(-1)` the non-contiguous color slice up front) is correct and worth keeping — only the "first axis" label is wrong.

## P3 — "~4–5× faster" overstates the CPU win

Measured on numpy 2.3.4 (best of 5): ~1.7–2.2× on greyscale (12-bit and full-range, 2000² and 6000²) and ~1.4× on the color slice — a real win, but not 4–5×, and it narrows toward parity as images grow. Soften to ~2× / "modestly faster (one `bincount` pass vs an introselect partition)", or drop the multiplier. The memory claim — the one this change is really about — holds; this is the side-benefit line only.

Checked: percentile math vs numpy's linear method (tests green, 32 cases), peak transient on greyscale / color-slice / wide-row paths, CPU across sizes and distributions, and description / CHANGELOG / comment consistency. The CHANGELOG (`#4968`, "large images", "output unchanged") and the `_HIST_BLOCK` comment are accurate and scoped; Greptile's `#????` note is stale (already filled).

🤖 Drafted with Claude; findings vetted by the reviewer — sanity-check rather than take as authoritative.

Reviewed against 0102e4d2
